### PR TITLE
Delay finding and associating a build file with a (handle) uri

### PR DESCRIPTION
**Pull request by nullptrdevs**
Current logic requires the modules available in a build.zig file to be known in order to correctly associate the file with an uri.

See #1629 

Take N2

### CODE REVIEW

File: src/DocumentStore.zig [hunk 1]
```diff
@@ -1087,6 +1083,24 @@ fn createDocument(self: *DocumentStore, uri: Uri, text: [:0]const u8, open: bool
...
+        });
+    }
+
+    return handle;
+}
+
+pub fn findBuildFileForUri(self: *DocumentStore, uri: Uri) error{OutOfMemory}!void {
+    const handle = self.getHandle(uri) orelse return;
     if (isBuildFile(handle.uri) and !isInStd(handle.uri)) {
         _ = self.getOrLoadBuildFile(handle.uri);
```
Comment (Techatrix) [inline src/DocumentStore.zig:1105]: This is unrelated to finding the `associated_build_file` so why does this happen here?

File: src/DocumentStore.zig [hunk 2]
```diff
@@ -1087,6 +1083,24 @@ fn createDocument(self: *DocumentStore, uri: Uri, text: [:0]const u8, open: bool
...
+        const server = @fieldParentPtr(Server, "document_store", self);
+
+        server.job_queue_lock.lock();
+        defer server.job_queue_lock.unlock();
+
+        try server.job_queue.ensureUnusedCapacity(1);
+        server.job_queue.writeItemAssumeCapacity(.{
+            .find_build_file_for_uri = try server.allocator.dupe(u8, uri),
+        });
+    }
```
Comment (Techatrix) [inline src/DocumentStore.zig:1097]: You have also delayed the collection imports and cimports which can cause wrong results until `findBuildFileForUri` is done. 

The code that prints the `Opened document {s} with build file {s}` message should also be removed because there is no guarantee that at that point in time, the associated build file has been found.

File: src/DocumentStore.zig [hunk 3]
```diff
@@ -1087,6 +1083,24 @@ fn createDocument(self: *DocumentStore, uri: Uri, text: [:0]const u8, open: bool
...
+
+    return handle;
+}
+
+pub fn findBuildFileForUri(self: *DocumentStore, uri: Uri) error{OutOfMemory}!void {
```
Comment (Techatrix) [inline src/DocumentStore.zig:1102]: This function potentially writes to `associated_build_file`, `import_uris` and `cimports` of `Handle` while other threads may read from them -> Data Race

**Review by Techatrix — CHANGES_REQUESTED** [review]

**Comment by Techatrix** [comment]
I just realized that you had converted your PR into a Draft, sorry for doing the review too early.

**Comment by nullptrdevs** [comment]
> I just realized that you had converted your PR into a Draft, sorry for doing the review too early.

It's actually appreciated — thank you.

I'm scrapping (the approach on) this one and could use ideas on how to solve it.

Haven't tested but I thought about refactoring out the find build file logic into a fn (similar to the one in this PR) and if there are no matches here
https://github.com/zigtools/zls/blob/b3285004f11b0a5f632a778685dcc8c1fcf50cf9/src/DocumentStore.zig#L1420-L1424
call it again? Con: extra file system operations

or 

Keep a `handle.potential_build_zig_files: []Uri` and check those.

File: src/DocumentStore.zig [hunk 4]
```diff
@@ -50,6 +50,12 @@ pub const BuildFile = struct {
         config: ?std.json.Parsed(BuildConfig) = null,
     } = .{},
 
+    pub fn hasConfig(self: *BuildFile) bool {
+        self.impl.mutex.lock();
+        defer self.impl.mutex.unlock();
+        return if (self.impl.config) |_| true else false;
```
Comment (Techatrix) [inline src/DocumentStore.zig:56]:
```suggestion
        return self.impl.config != null;
```

File: src/DocumentStore.zig [hunk 5]
```diff
@@ -588,6 +600,32 @@ fn getOrLoadBuildFile(self: *DocumentStore, uri: Uri) ?*BuildFile {
...
+            handle.associated_build_file = handle.potential_build_file_uris.items[0];
+            if (count == 1) break :blk;
+            const ok: bool = ok: for (handle.potential_build_file_uris.items[1..], 1..) |build_file_uri, index| {
+                const build_file = self.getOrLoadBuildFile(build_file_uri) orelse break :ok false;
+                if (!build_file.hasConfig()) break :ok false;
+                if (try self.uriAssociatedWithBuild(build_file, handle.uri)) {
+                    handle.associated_build_file = handle.potential_build_file_uris.items[index];
+                    break :ok true;
+                }
+            } else unreachable;
```
Comment (Techatrix) [inline src/DocumentStore.zig:619]: Why is the else `unreachable`?

File: src/DocumentStore.zig [hunk 6]
```diff
@@ -588,6 +600,32 @@ fn getOrLoadBuildFile(self: *DocumentStore, uri: Uri) ?*BuildFile {
...
+pub fn doubleCheckPotentialBuildFileUri(self: *DocumentStore, handle: *Handle) error{OutOfMemory}!?Uri {
+    handle.impl.lock.lock();
+    defer handle.impl.lock.unlock();
+    switch (handle.potential_build_file_uris.items.len) {
+        0 => {},
+        else => |count| blk: {
+            _ = self.getOrLoadBuildFile(handle.potential_build_file_uris.items[0]) orelse break :blk;
+            handle.associated_build_file = handle.potential_build_file_uris.items[0];
+            if (count == 1) break :blk;
+            const ok: bool = ok: for (handle.potential_build_file_uris.items[1..], 1..) |build_file_uri, index| {
```
Comment (Techatrix) [inline src/DocumentStore.zig:612]: How about have a loop with the following structure?
`handle.associated_build_file = for (handle.potential_build_file_uris.items) { ... };`

You can have a special check before for `handle.potential_build_file_uris.items.len == 1`. This should avoid having the boolean to reset the associated_build_file after the loop.

File: src/DocumentStore.zig [hunk 7]
```diff
@@ -588,6 +600,32 @@ fn getOrLoadBuildFile(self: *DocumentStore, uri: Uri) ?*BuildFile {
     return gop.value_ptr.*;
 }
 
+pub fn doubleCheckPotentialBuildFileUri(self: *DocumentStore, handle: *Handle) error{OutOfMemory}!?Uri {
```
Comment (Techatrix) [inline src/DocumentStore.zig:603]: If we have a handle with multiple potential build files but none of them are associated to the handle, then this functions returns `null`. Every subsequent call will then do its entire work again and then still return `null`. There should be some kind of caching. This should also help with reducing contention on the handle mutex.

**Review by Techatrix — CHANGES_REQUESTED** [review]

File: src/DocumentStore.zig [hunk 8]
```diff
@@ -1063,7 +1101,7 @@ fn uriInImports(
     };
     gop.key_ptr.* = handle.uri;
 
-    if (handle.associated_build_file) |associated_build_file_uri| {
+    if (handle.associated_build_file orelse try self.doubleCheckPotentialBuildFileUri(handle)) |associated_build_file_uri| {
```
[the remark this answers is not in the document]
Comment (nullptrdevs) [reply src/DocumentStore.zig:1104]: Within that fn, is a `return self.associated_build_file;` guaranteed to make a copy or should the lock be held outside?

File: src/DocumentStore.zig (hunk 7 above)
Comment (nullptrdevs) [reply src/DocumentStore.zig:603]: I'd say the issue is when there are none, otherwise if they all've been ran the first one sticks.

I'm thinking maybe passing an optional handle uri to getOrLoadBuildFile to do the association after the file has been ran (for every build file other than the first one). It'd be easier to pass a *Handle, but having an uri and calling getHandle on it ensures the handle is still valid.

File: src/DocumentStore.zig (hunk 5 above)
Comment (nullptrdevs) [reply src/DocumentStore.zig:619]: The `for` is guaranteed at least one entry/iteration.

Comment (Techatrix) [reply src/DocumentStore.zig:619]: How is the number of iterations related to whether or the not the else path is reached?

You code is equivalent to:
```zig
for (handle.potential_build_file_uris.items[1..], 1..) |build_file_uri, index| {
    const build_file = self.getOrLoadBuildFile(build_file_uri) orelse break :ok false;
    if (!build_file.hasConfig()) break :ok false;
    if (try self.uriAssociatedWithBuild(build_file, handle.uri)) {
        handle.associated_build_file = handle.potential_build_file_uris.items[index];
        break :ok true;
    }
    // if this code path here is reached on the last iteration the you hit unreachable.
    // I don't see why it shouldn't be possible to reach on the last iteration.
    // The handle may not be associated with any of the potential build files.
} else unreachable
```

File: src/DocumentStore.zig (hunk 8 above)
[the remark this answers is not in the document]
Comment (Techatrix) [reply src/DocumentStore.zig:1104]: Within the function, you are holding a lock so the read from `self.associated_build_file` is protected but outside the function, you are not holding a lock which makes accessing `self.associated_build_file` a potential data race.
In short, always hold the lock of the handle when accessing `self.associated_build_file`.

Does that make it more clear what the problem is?

File: src/DocumentStore.zig (hunk 7 above)
Comment (Techatrix) [reply src/DocumentStore.zig:603]: Isn't it possible to return null when searching through multiple potential build files because of the `if (!ok) handle.associated_build_file = null;` line? If `!ok` then we return `null` and any subsequent call would also return `null`.

File: src/DocumentStore.zig (hunk 5 above)
Comment (nullptrdevs) [reply src/DocumentStore.zig:619]: TIL; I thought it'd be only if the condition failed.

File: src/DocumentStore.zig (hunk 8 above)
[the remark this answers is not in the document]
Comment (nullptrdevs) [reply src/DocumentStore.zig:1104]: I get the data race, my question is if  the `handle.getAssociatedBuildFile(store)` is
```zig
// lock defer unlock here
return self.associated_build_file;
```
is it guaranteed to make a copy which can be safely read outside the fn or should it be similar to`BuildFile.tryLockConfig`

File: src/DocumentStore.zig (hunk 7 above)
Comment (nullptrdevs) [reply src/DocumentStore.zig:603]: Equipped with the knowledge of how `for() {} else` works I think correcting the else to `else true;` should handle this condition;

**Comment by nullptrdevs** [comment]
Did the 'pass an optional handle uri to getOrLoadBuildFile to do the association after the file has been ran', but it had a blind spot where if you had a 'build.zig' file open it would be created without the "callback", would get loaded, but have no config when checked via the "callback", so a recheck in `handle.getAssociatedBuildFileUri` was needed.

Should we keep the "callback"  complexity or have the "on demand" only logic?

**Review by Techatrix — CHANGES_REQUESTED** [review]
> Should we keep the "callback" complexity or have the "on demand" only logic?

I would consider your current "callback" approach to be acceptable (but a bit complex). I you want to do the "on demand" approach then this implementation of `getAssociatedBuildFileUri` could be feasible.

Keep in mind that I have not tested this code and that TODO should definitely be resolved.
```zig
pub fn getAssociatedBuildFileUri(self: *Handle, store: *DocumentStore) ?Uri {
    self.impl.lock.lock();
    defer self.impl.lock.unlock();
    switch(associated_build_file) {
        .none => return null,
        .unresolved => |potential_build_file_uris| {
            // should there be special logic if `potential_build_file_uris.len == 1`?
            
            var has_missing_build_config = false;
            for (potential_build_file_uris) |build_file_uri| {
                const build_file = self.getOrLoadBuildFile(build_file_uri, handle.uri) orelse continue; // continue or break?
                if (!build_file.hasConfig()) {
                    has_missing_build_config = true;
                    continue;
                }
                if (!try self.uriAssociatedWithBuild(build_file, handle.uri)) {
                    // TODO
                    // the given `build_file_uri` should be removed from `potential_build_file_uris`
                    // so that future calls don't redo all this work on every call.
                    continue;
                }
                /// the associated build file has been resolved so there is no need to store the potential build files.
                handle.potential_build_file_uris.clearAndFree(handle.allocator);
    
                associateBuildFileUri = .{ .resolved = build_file.uri };
                return build_file.uri;
            };
    
            if (!has_missing_build_config) {
                /// there is no associated build file so there is no need to store the potential build files.
                handle.potential_build_file_uris.clearAndFree(handle.allocator);
                associated_build_file = .none;
            }
            // when build configs are missing we keep the state at .unresolved so that
            // future calls will retry until all build config are resolved.
            // Then will have a conclusive result on whether or not there is a associated build file.
            return null;
        },
        .resolved => |uri| return uri,
    }
}
```
```zig
associated_build_file: union(enum) {
    /// The Handle has no associated build file (build.zig).
    none,
    /// The associated build file (build.zig) has not been resolved yet.
    /// Stores a list of potential build files.
    unresolved: []const Uri,
    /// The associated build file (build.zig) has been successfully resolved.
    resolved: Uri,
} = .unresolved,
```

File: src/DocumentStore.zig [hunk 9]
```diff
@@ -1072,10 +1124,6 @@ fn uriInImports(
     };
     gop.key_ptr.* = handle.uri;
 
-    if (handle.associated_build_file) |associated_build_file_uri| {
-        return std.mem.eql(u8, associated_build_file_uri, build_file_uri);
-    }
```
[the remark this answers is not in the document]
Comment (nullptrdevs) [reply src/DocumentStore.zig:1077]: Noice - a clear objective! ..also, I put those back in :sweat_smile:, which restored same timings as `main`.
..but it was still slow (on my machine/fs), ..so I took it further :zany_face:

File: src/DocumentStore.zig [hunk 10]
```diff
@@ -360,6 +363,26 @@ pub const Handle = struct {
         return @bitCast(self.impl.status.load(.Acquire));
     }
 
+    /// Non-blocking - tries to obtain a shared lock,
+    /// holds a `shared` lock on success
+    /// Release with `handle.unlockShared();`
+    pub fn getAssociatedBuildFileUri(self: *Handle, store: *DocumentStore, associate_if_pending: bool) ?Uri {
+        if (!self.impl.lock.tryLockShared()) return null;
```
[the remark this answers is not in the document]
Comment (nullptrdevs) [reply src/DocumentStore.zig:370]: Most of the calls are for completions, so I didn't consider it necessary.

File: src/DocumentStore.zig [hunk 11]
```diff
@@ -483,9 +509,10 @@ allocator: std.mem.Allocator,
 config: *const Config,
 /// the DocumentStore assumes that `runtime_zig_version` is not modified while calling one of its functions.
 runtime_zig_version: *const ?ZigVersionWrapper,
-lock: std.Thread.RwLock = .{},
+lock: RwLock = .{},
 handles: std.StringArrayHashMapUnmanaged(*Handle) = .{},
 build_files: std.StringArrayHashMapUnmanaged(*BuildFile) = .{},
+build_files_lock: RwLock = .{},
```
[the remark this answers is not in the document]
Comment (nullptrdevs) [reply src/DocumentStore.zig:515]: I favor granularity, but can't think of a scenario where contention will be too high, .. still

File: src/DocumentStore.zig [hunk 12]
```diff
@@ -594,6 +622,23 @@ fn getOrLoadBuildFile(self: *DocumentStore, uri: Uri) ?*BuildFile {
...
 
+pub fn associateBuildFileUri(self: *DocumentStore, handle: *Handle) error{OutOfMemory}!void {
+    handle.impl.lock.lock();
+    defer handle.impl.lock.unlock();
+
+    if (handle.impl.associated_build_file != .pending) return;
+
+    handle.impl.associated_build_file = blk: for (handle.potential_build_file_uris.items, 0..) |build_file_uri, index| {
+        const build_file = self.getOrLoadBuildFile(build_file_uri, handle.uri) orelse break :blk .pending;
+        if (!build_file.hasConfig()) break :blk .pending;
```
[the remark this answers is not in the document]
Comment (nullptrdevs) [reply src/DocumentStore.zig:633]: But then it would reach the `else => .none`, unless I come around and implement your suggestion

File: src/DocumentStore.zig [hunk 13]
```diff
@@ -594,6 +622,23 @@ fn getOrLoadBuildFile(self: *DocumentStore, uri: Uri) ?*BuildFile {
...
+        if (try self.uriAssociatedWithBuild(build_file, handle.uri)) break :blk .{ .uri = handle.potential_build_file_uris.items[index] };
+    } else .none;
```
[the remark this answers is not in the document]
Comment (nullptrdevs) [reply src/DocumentStore.zig:635]: I'm still to implement the `TODO` but I'm thinking
```zig
pending: struct {
	potential: []const Uri,
    last_checked_index: usize,
};
```
to avoid extra allocations?

File: src/DocumentStore.zig [hunk 14]
```diff
@@ -35,31 +35,45 @@ pub fn computeHash(bytes: []const u8) Hash {
...
     } = .{},
 
-    pub fn tryLockConfig(self: *BuildFile) ?BuildConfig {
-        self.impl.mutex.lock();
-        return if (self.impl.config) |cfg| cfg.value else {
-            self.impl.mutex.unlock();
-            return null;
-        };
+    pub fn hasConfig(self: *BuildFile) bool {
+        if (!self.impl.lock.tryLockShared()) return false;
```
[the remark this answers is not in the document]
Comment (nullptrdevs) [reply src/DocumentStore.zig:59]: I'd like to keep them - technically, they'd succeed 99.9% of the time and more an of indication that it isn't critical.
